local-cluster/tests/local_cluster_slow_1.rs

```diff
@@ -386,10 +391,360 @@ fn test_kill_partition_switch_threshold_progress() {
+    assert_eq!(dup_shred1.slot(), dup_shred2.slot());
+    assert_eq!(dup_shred1.slot(), dup_slot);
+
+    // Purge everything including the `dup_slot` from the `target_switch_fork_validator_pubkey`
+    info!("Purging towers and ledgers of duplicate slot: {}", dup_slot);
+    clear_ledger_and_tower(
+        &target_switch_fork_validator_ledger_path,
+        &target_switch_fork_validator_pubkey,
+        dup_slot,
+    );
+    clear_ledger_and_tower(
+        &duplicate_fork_validator1_ledger_path,
+        &duplicate_fork_validator1_pubkey,
+        dup_slot + 1,
+    );
+    clear_ledger_and_tower(
+        &duplicate_fork_validator2_ledger_path,
+        &duplicate_fork_validator2_pubkey,
+        dup_slot + 1,
+    );
+
+    // Set entrypoint to `target_switch_fork_validator_pubkey` so we can run discovery in gossip even without the
+    // bad leader
+    cluster.set_entry_point(target_switch_fork_validator_info.info.contact_info.clone());
+
+    // 4) Restart `target_switch_fork_validator_pubkey`, and ensure they vote on their own leader slot
+    // not descended from the duplicate slot
+    info!("Restarting switch fork node");
+    cluster.restart_node(
+        &target_switch_fork_validator_pubkey,
+        target_switch_fork_validator_info,
+        SocketAddrSpace::Unspecified,
+    );
+    let target_switch_fork_validator_ledger_path =
+        cluster.ledger_path(&target_switch_fork_validator_pubkey);
+
+    info!("Waiting for switch fork to make block past duplicate fork");
+    loop {
+        let last_vote = last_vote_in_tower(
+            &target_switch_fork_validator_ledger_path,
+            &target_switch_fork_validator_pubkey,
+        );
+        if let Some((latest_vote_slot, _hash)) = last_vote {
+            if latest_vote_slot > dup_slot {
+                let blockstore = open_blockstore(&target_switch_fork_validator_ledger_path);
+                let ancestor_slots: HashSet<Slot> =
+                    AncestorIterator::new_inclusive(latest_vote_slot, &blockstore).collect();
+                assert!(ancestor_slots.contains(&latest_vote_slot));
+                assert!(ancestor_slots.contains(&0));
+                assert!(!ancestor_slots.contains(&dup_slot));
+                break;
+            }
+        }
+        sleep(Duration::from_millis(1000));
+    }
+
+    // Now restart the duplicate validators
+    // Start the node with partition enabled so they don't see the `target_switch_fork_validator_pubkey`
+    // before voting on the duplicate block
+    info!("Restarting duplicate fork node");
+    // Ensure `duplicate_fork_validator1_pubkey` votes before starting up `duplicate_fork_validator2_pubkey`
+    // to prevent them seeing `dup_slot` as duplicate confirmed before voting.
+    restart_dup_validator(
+        &mut cluster,
+        duplicate_fork_validator1_info,
+        &duplicate_fork_validator1_pubkey,
+        dup_slot,
+        &dup_shred1,
+        &dup_shred2,
+    );
```

Comment:
Is there anyway to guarantee that we don't race here by making sure `duplicate_fork_validator1_pubkey1` is attempting to switch forks *before* it observes `duplicate_fork_validator2_pubkey` vote? If it happens in quick succession we might not actually test the reset logic.

Reply:
Maybe just check to see that it's PoH is properly reset here?

Reply:
hmmm, so to go over again why the switch fork is necessary:

So the events go:

1. Vote for duplicate block `D`
2. See `D` is duplicate, remove from fork choice and reset to ancestor `A`, potentially generating a fork off that ancestor
3. See `D` is duplicate confirmed, but because of the bug where we disallow resetting to a slot which matches the last vote slot, we still don't build off `D`, and continue building on `A`.

So the switch fork is necessary in 2. to prevent the validator from making a freebie vote  from `A` and allowing consensus to continue.

So I think you're right, we should make sure:

It sees the target switch fork *before* it can make another vote on any forks it generates from `A`. The prevention of forks from `A` is kind of tricky, I kind of tried to hack it with this 

```
 let validator_to_slots = vec![
        (duplicate_leader_validator_pubkey, 80),
        (target_switch_fork_validator_pubkey, 80),
        // Give the `duplicate_fork_validator1_pubkey` very few leader slots so we have ample time to:
        // 1. Give him a duplicate version of the slot
        // 2. Allow him to vote on that duplicate slot
        // 3. See that the slot was duplicate AFTER he voted
        //
        // Only at this time do we want this validator to have a chance to build a block. Otherwise,
        // if the validator has opportunity to build a block between 2 and 3, they'll build the block
        // on top of the duplicate block, which will possibly includeh is vote for the duplicate block. We
        // want to avoid this because this will make fork choice pick the duplicate block.
        (duplicate_fork_validator1_pubkey, 4),
    ];
```

which doesn't give the `duplicate_fork_validator1_pubkey` many leader slots

Reply:
I'll add a comment to the code explaining this

Reply:
yeah makes sense